src/hooks/useConfig.ts

```diff
@@ -0,0 +1,50 @@
+import {useState} from 'react';
+
+import {Config, MidiInputConfig} from '../types/config-types';
+import {mockMidiInputConfig} from '../mockfiles/MockMidiDevice';
+
+export type UseConfigHookValue = {
+  config: Config;
+
+  setConfig: (config: Config) => void;
+  updateConfigForInputDevice: (inputConfig: MidiInputConfig) => void;
+}
+
+const initialConfig: Config = {
+  midi: {
+    inputs: [
+      mockMidiInputConfig,
+    ],
+    outputs: [],
+  }
+}
+
+export const useConfig = (): UseConfigHookValue => {
+  const [config, setConfig] = useState<Config>(initialConfig);
+
+  const updateConfigForInputDevice = (inputConfig: MidiInputConfig) => {
+    const inputConfigs = config.midi.inputs;
+
+    const index = inputConfigs.findIndex(input => input.name === inputConfig.name);
+    let newConfigState: MidiInputConfig[];
+    if (index === -1) {
+      newConfigState = [...inputConfigs, inputConfig];
+    } else {
+      newConfigState = [...inputConfigs.slice(0, index), inputConfig, ...inputConfigs.slice(index + 1)];
+    }
+
+    setConfig({
+      ...config,
+      midi: {
+        ...config.midi,
+        inputs: newConfigState,
+      },
+    });
+  };
```

Comment:
Here we're using the "spread" syntax to create a new object, based on properties of an existing object. This is the equivalent of:

```ts
config.midi.inputs = newConfigState;
setConfig(config);
```

except we are creating a new `config`, instead of mutating the existing config. This is so React can use a new reference of the entire config in the `config` state value. There are libraries that allow you to do the mutation syntax and still end up with "immutable" data https://github.com/immerjs/immer